src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,520 @@
+				        text,
+				        this_fn
+				);
+			}
+			break;
+		}
+		case UI_SCRIPT_INVALID_BEFORE:
+		case UI_SCRIPT_INVALID_HEREAFTER:
+			ui_displayUint64Screen(
+			        HEADER,
+			        ctx->scriptContent.timelock,
+			        this_fn
+			);
+			break;
+		default:
+			THROW(ERR_INVALID_STATE);
+		}
+	}
+
+	UI_STEP(DISPLAY_UI_STEP_RESPOND) {
+		io_send_buf(SUCCESS, NULL, 0);
+		ui_displayBusy();
+	}
+
+	UI_STEP_END(DISPLAY_UI_STEP_INVALID);
+}
+#undef HEADER
+#undef ASSERT_UI_SCRIPT_TYPE_SANITY
+
+#define UI_DISPLAY_SCRIPT(UI_TYPE)\
+	ctx->ui_scriptType = UI_TYPE;\
+	ctx->ui_step = DISPLAY_UI_STEP_POSITION;\
+	deriveScriptHash_display_ui_runStep();
+
+// Start complex native script
+
+static void deriveNativeScriptHash_handleAll(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_all(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ALL;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ALL);
+}
+
+static void deriveNativeScriptHash_handleAny(read_view_t* view)
+{
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("");
+
+	nativeScriptHashBuilder_startComplexScript_any(&ctx->hashBuilder, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_ANY;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_ANY);
+}
+
+static void deriveNativeScriptHash_handleNofK(read_view_t* view)
+{
+	// parse data
+	ctx->scriptContent.requiredScripts = parse_u4be(view);
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE_WITH_CTX("required scripts = %u, ", ctx->scriptContent.requiredScripts);
+
+	// validate that the received requiredScripts count makes sense
+	VALIDATE(ctx->complexScripts[ctx->level].remainingScripts >= ctx->scriptContent.requiredScripts, ERR_INVALID_DATA);
+
+	nativeScriptHashBuilder_startComplexScript_n_of_k(&ctx->hashBuilder, ctx->scriptContent.requiredScripts, ctx->complexScripts[ctx->level].remainingScripts);
+
+	ctx->complexScripts[ctx->level].ui_type = UI_SCRIPT_N_OF_K;
+	UI_DISPLAY_SCRIPT(UI_SCRIPT_N_OF_K);
+}
+
+static void deriveNativeScriptHash_handleComplexScriptStart(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	expectsMoreScripts();
+
+	// check if we can increase the level without breaking the MAX_SCRIPT_DEPTH constraint
+	VALIDATE(ctx->level + 1 < MAX_SCRIPT_DEPTH, ERR_DATA_TOO_LARGE);
+	ctx->level++;
+	explicit_bzero(&ctx->complexScripts[ctx->level], SIZEOF(ctx->complexScripts[ctx->level]));
+
+	uint8_t nativeScriptType = parse_u1be(view);
+	TRACE("native complex script type = %u", nativeScriptType);
+
+	switch(nativeScriptType) {
+#	define  CASE(TYPE, HANDLER) case TYPE:\
+		ctx->complexScripts[ctx->level].remainingScripts = parse_u4be(view);\
+		ctx->complexScripts[ctx->level].totalScripts = ctx->complexScripts[ctx->level].remainingScripts;\
```

Comment:
These two lines are independent of the CASE macro arguments, that's very fishy. Can't they be moved before the switch?

Reply:
if it doesn't matter if I parse the other argument before I know the first one is valid I will fix this. Also I think there are other instances where I do something similar so I will also fix those.

Reply:
Hm. But it increases code size, which is annoying in the context of Ledger. Let's move them before the switch and add a comment to 
```
	uint8_t nativeScriptType = parse_u1be(view);
	TRACE("native complex script type = %u", nativeScriptType);
```
that the value is validated below.